File: 129.Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root, prefix=0):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return prefix
        prefix = root.val + prefix * 10
        if root.left is None and root.right is None:
            return prefix
        if root.left is None or root.right is None:
            return self.sumNumbers(root.left or root.right, prefix)
        return self.sumNumbers(root.left, prefix) + self.sumNumbers(root.right, prefix)

File: 129.Sum Root to Leaf Numbers.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def sum_numbers(root, prefix=0)
    if root.nil?
        0
    elsif root.left.nil? && root.right.nil?
        prefix * 10 + root.val
    else
        sum_numbers(root.left, prefix * 10 + root.val) + sum_numbers(root.right, prefix * 10 + root.val)
    end
end
